crates/exec-wasmtime/src/runtime/vfs/listen.rs

```diff
@@ -0,0 +1,163 @@
+    ) -> WasiResult<Box<dyn WasiFile>> {
+        if dir {
+            return Err(Error::not_dir());
+        }
+
+        if !read || !write {
+            return Err(Error::perm()); // FIXME(@npmccallum): errno
+        }
+
+        let addr = path
+            .rsplit_terminator(MAIN_SEPARATOR)
+            .next()
+            .ok_or_else(|| Error::invalid_argument().context("failed to parse file name"))?;
+        let (host, port) = addr
+            .split_once(':')
+            .map(|(host, addr)| (Some(host), addr))
+            .unwrap_or((None, addr));
+        let port = port
+            .parse()
+            .map_err(|e| Error::invalid_argument().context(e))
+            .context("failed to parse port `{port}`")?;
+        let tcp = match (host, port) {
+            (Some("localhost" | "127.0.0.1"), port) => std::net::TcpListener::bind(SocketAddr::V4(
+                SocketAddrV4::new(Ipv4Addr::LOCALHOST, port),
+            )),
+            (None | Some("0.0.0.0"), port) => std::net::TcpListener::bind(SocketAddr::V4(
+                SocketAddrV4::new(Ipv4Addr::UNSPECIFIED, port),
+            )),
+            (Some(host), port) => std::net::TcpListener::bind((host, port)),
+        }
+        .map(TcpListener::from_std)
+        .map_err(|e| Error::io().context(e).context("failed to bind to socket"))?;
+
+        if flags == FdFlags::NONBLOCK {
+            tcp.set_nonblocking(true)
+                .context("failed to enable NONBLOCK")?;
+        } else if flags.is_empty() {
+            tcp.set_nonblocking(false)
+                .context("failed to disable NONBLOCK")?;
+        } else {
+            return Err(
+                Error::invalid_argument().context("cannot set anything other than NONBLOCK")
+            );
+        }
+        match self.as_ref() {
+            Self::Tcp(..) => Ok(wasmtime_wasi::net::Socket::from(tcp).into()),
+            Self::Tls(Link { inode, .. }) => {
+                Ok(tls::Listener::new(tcp, inode.data.read().await.content.clone()).into())
+            }
+        }
+    }
+}
+pub async fn new(
+    parent: Arc<dyn Node>,
+    certs: Arc<Vec<Certificate>>,
+    key: Arc<Zeroizing<Vec<u8>>>,
+    sockets: impl IntoIterator<Item = (FileName, ListenFile)>,
+) -> anyhow::Result<Arc<dyn Node>> {
+    let certs = certs.deref().clone();
+    let key = PrivateKey(key.deref().deref().clone());
+    let tls_config = rustls::ServerConfig::builder()
+        .with_cipher_suites(tls::DEFAULT_CIPHER_SUITES.deref())
+        .with_kx_groups(tls::DEFAULT_KX_GROUPS.deref())
+        .with_protocol_versions(tls::DEFAULT_PROTOCOL_VERSIONS.deref())?
+        .with_no_client_auth() // TODO: https://github.com/enarx/enarx/issues/1547
+        .with_single_cert(certs, key)
+        .map(Arc::new)
+        .context("failed to construct TLS config")?;
+    let dir = Directory::device(parent, {
+        let tls_config = tls_config.clone();
+        Some(Arc::new(move |parent: Arc<dyn Node>| {
```

Comment:
Why are we `Arc::new`ing a closure? Where is this closure invoked?

Reply:
See VFS crate source. This creates a new file in the directory https://github.com/enarx/vfs/pull/18